Sources/Leaf/Renderer/Renderer.swift

```diff
@@ -0,0 +1,69 @@
+import Bits
+
+/// Renders Leaf templates using the Leaf parser and serializer.
+public final class Renderer {
+    /// The tags available to this renderer.
+    public let tags: [String: Tag]
+
+    /// The renderer will use this to read files for
+    /// tags that require it (such as #embed)
+    public let fileReader: FileReader
+
+    /// Create a new Leaf renderer.
+    public init(tags: [String: Tag]? = nil, fileReader: FileReader) {
+        self.tags = tags ?? defaultTags
+        self.fileReader = fileReader
+    }
+
+    // ASTs only need to be parsed once
+    private var _cachedASTs: [Int: [Syntax]] = [:]
+
+    /// Renders the supplied template bytes into a view
+    /// using the supplied context.
+    public func render(_ template: Bytes, context: DataRepresentable) throws -> Bytes {
+        let hash = template.makeString().hashValue
+
+        let ast: [Syntax]
+        if let cached = _cachedASTs[hash] {
+            ast = cached
+        } else {
+            let parser = Parser(template)
+            do {
+                ast = try parser.parse()
+            } catch let error as ParserError {
+                throw RenderError(source: error.source, reason: error.reason, error: error)
+            }
+            _cachedASTs[hash] = ast
```

Comment:
String compares are O(n^2) in complexity which given the size of templates would make this pretty slow. Originally we were doing caching based on file path, but that has the issue that Leaf will continue to cache a file even if its changed. 

Though now that I think about it, with this method files must still be read from disk (and hashed) w/ each request which is probably significantly slower. Reverting to the original method with the optional cache passed to the `Renderer` is probably the best way to go.